server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java

```diff
@@ -33,14 +39,34 @@
 
    private final HotRodServer server;
    private final NettyTransport transport;
+   private final Executor executor;
 
-   public ContextHandler(HotRodServer server, NettyTransport transport) {
+   public ContextHandler(HotRodServer server, NettyTransport transport, Executor executor) {
       this.server = server;
       this.transport = transport;
+      this.executor = executor;
    }
 
    @Override
    protected void channelRead0(ChannelHandlerContext ctx, CacheDecodeContext msg) throws Exception {
+      executor.execute(() -> {
+         try {
+            Subject subject = msg.getSubject();
+            if (subject == null)
+               realRead(ctx, msg);
+            else Security.doAs(subject, (PrivilegedExceptionAction<Void>) () -> {
```

Comment:
Security checks would be per-cache while the transport is per-container. So you'd need to check anyway